Update URL tests

domenic opened the pull request:
The current URL tests were not up to date with the latest URL spec, especially with regard to file URLs. In addition, they did not test serialization against predetermined results, but instead implemented an incorrect version of the URL Standard's algorithm (this is #2038). Finally, they left a lot of things implicit.

This moves to a new test data format that is explicit for every component of the URL, for maximum coverage. This includes previously-untested components like origin, username, and password.

R=@annevk I assume

domenic committed:
Update URL tests format

This moves to a new test data format that is explicit for most components of the URL, to avoid logic in the tests, fixing #2038.

tobie commented:
Reviewers for this pull request are: @w3c/url-reviewers.

zcorpan commented on url/a-element.html:
```diff
@@ -9,11 +9,11 @@
-  request.open("GET", "urltestdata.txt")
+  request.open("GET", "urltestdata.json")
   request.send()
   request.responseType = "text"
```
Use `responseType = "json"` instead?

domenic replied:
I wasn't sure it was a good ideal to couple to json responseType support. But I guess all modern browsers have that.

domenic replied:
Fixed

zcorpan commented on url/urltestdata.txt:
```diff
@@ -1,402 +0,0 @@
-# FORMAT NOT DOCUMENTED YET (parser is urltestparser.js)
-
-# Based on http://trac.webkit.org/browser/trunk/LayoutTests/fast/url/script-tests/segments.js
```
Can we retain the comments somehow? (e.g. as strings between tests)

annevk replied:
That or maybe in a README. They have been changed a bit too so the reference is no longer that valuable.

domenic replied:
For now I've restored the comments, except for the ones that were obviously wrong (saying "this should fail" when due to previous updates they no longer fail). Some of them are reasonably valuable, talking about why something fails.

zcorpan commented:
I haven't really reviewed the changes to the tests themselves, but otherwise LGTM. (I would have been easier to review changes to the tests if it was converted to json in one commit and then fixes in another.)

domenic committed:
Update URL tests

The URL tests were not up to date with the latest URL spec, especially with regard to file URLs.

domenic committed:
Add to URL tests

Add back the comments that were originally in urltestdata; add tests for origin, username, password, and host components; add a few early tests from the jsdom/whatwg-url repo.

domenic commented:
@zcorpan @annevk I separated this out into three commits which should make it easier to review.

- The first simply updates to JSON format with no change in test data.
- The second updates the tests for the latest URL Standard, at least according to our reference implementation. This is probably where the closest review is needed.
- The third adds extra tests and tests for the origin, username, password, and host components. It also restores the comments because it was easiest for me to do that at the end instead of trying to preserve them through the diffs.

zcorpan commented:
> The second updates the tests for the latest URL Standard, at least according to our reference implementation. This is probably where the closest review is needed.

Reviewed, LGTM % @annevk's comment.

domenic commented:
Updated to mark that one back to failure; filed https://bugs.chromium.org/p/v8/issues/detail?id=4654 to fix Chrome.
